src/hvmc_net/mutual_recursion.rs

```diff
@@ -0,0 +1,130 @@
+use hvmc::ast::{Book, Tree};
+use indexmap::{IndexMap, IndexSet};
+use std::fmt::Debug;
+
+type Ref = String;
+type Stack<T> = Vec<T>;
+type RefSet = IndexSet<Ref>;
+
+#[derive(Default)]
+pub struct Graph(IndexMap<Ref, RefSet>);
+
+pub fn check_cycles(book: &Book) -> Result<(), String> {
+  let graph = Graph::from(book);
+  let cycles = graph.cycles();
+
+  if cycles.is_empty() { Ok(()) } else { Err(pretty_print_cycles(&cycles)) }
+}
+
+fn pretty_print_cycles(cycles: &[Vec<Ref>]) -> String {
+  cycles
+    .iter()
+    .enumerate()
+    .map(|(i, cycles)| format!("Cycle {}: {}", 1 + i, cycles.join(" -> ")))
```

Comment:
We need to improve this message.
It should be something like `Mutual recursion cycle detected in compiled functions. Cycle 1: A -> B -> ... -> A. This program will expand infinitely in strict evaluation mode. Read <link to docs> for more information.`